Implement permutedims

kescobo opened the pull request:
Figured I'd get a jumpstart on Hacktoberfest, and this is something I've wanted in DataFrames for a longtime. This is an attempt to address #2420 

## todo
- [x] correct function signature
- [x] docs
- [x] tests

## So far

I implemented 3 versions of `transpose`:

- `transpose1` copies @tbeason 's stack/unstack method from the issue above
- `transpose2` creates a `Matrix` from everything _except_ the indexing column, and does `permutedims` on it
- `transpose3` builds column-by-column by iterating rows

## Benchmarking

```
julia> builddf(n) = DataFrame(a = [randstring(6) for _ in 1:n], b = rand(n), c = rand(Int, n), d = repeat(["x", "y"], inner=n÷2))
builddf (generic function with 1 method)

julia> df1 = builddf(10);

julia> df2 = builddf(100);

julia> df3 = builddf(1000);

julia> @btime DataFrames.transpose1($df1);
  33.565 μs (398 allocations: 32.61 KiB)

julia> @btime DataFrames.transpose2($df1);
  12.637 μs (179 allocations: 17.81 KiB)

julia> @btime DataFrames.transpose3($df1);
  14.723 μs (172 allocations: 16.58 KiB)

julia> @btime DataFrames.transpose1($df2);
  144.708 μs (1083 allocations: 115.92 KiB)

julia> @btime DataFrames.transpose2($df2);
  50.667 μs (650 allocations: 69.75 KiB)

julia> @btime DataFrames.transpose3($df2);
  137.353 μs (1261 allocations: 119.12 KiB)

julia> @btime DataFrames.transpose1($df3);
  1.168 ms (7924 allocations: 976.30 KiB)

julia> @btime DataFrames.transpose2($df3);
  469.847 μs (5678 allocations: 643.34 KiB)

julia> @btime DataFrames.transpose3($df3);
  1.513 ms (20035 allocations: 1.27 MiB)
```

## First impressions

- `transpose1` (stack/unstack) is slower on small `DataFrame`s but scales well.
     - This also has the possible disadvantage that the columns end up sorted, rather than preserving the row order
- `transpose2` is the fastest, but might be harder to fix type issues (see below)
- `transpose3` is pretty fast for small `DataFrame`s but doesn't scale well at all

## Other thoughts

All 3 versions end up with `Any` typed columns, regardless of what's in the resulting column (unless all the columns are the same type). This seems like it might be more easily fixed with 1 and 3. The behavior is also not quite what I was expecting if all of the columns are numerical:

```
julia> df_types = DataFrame(a=[:x, :y, :z], b=rand(3), c=["foo", 2.2, 1])
3×3 DataFrame
│ Row │ a      │ b        │ c   │
│     │ Symbol │ Float64  │ Any │
├─────┼────────┼──────────┼─────┤
│ 1   │ x      │ 0.633028 │ foo │
│ 2   │ y      │ 0.388354 │ 2.2 │
│ 3   │ z      │ 0.134275 │ 1   │

julia> DataFrames.transpose1(df_types)
2×4 DataFrame
│ Row │ variable │ x        │ y        │ z        │
│     │ String   │ Any      │ Any      │ Any      │
├─────┼──────────┼──────────┼──────────┼──────────┤
│ 1   │ b        │ 0.633028 │ 0.388354 │ 0.134275 │
│ 2   │ c        │ foo      │ 2.2      │ 1        │

julia> DataFrames.transpose2(df_types)
2×4 DataFrame
│ Row │ variable │ x        │ y        │ z        │
│     │ String   │ Any      │ Any      │ Any      │
├─────┼──────────┼──────────┼──────────┼──────────┤
│ 1   │ b        │ 0.633028 │ 0.388354 │ 0.134275 │
│ 2   │ c        │ foo      │ 2.2      │ 1        │

julia> DataFrames.transpose3(df_types)
2×4 DataFrame
│ Row │ variable │ x        │ y        │ z        │
│     │ String   │ Any      │ Any      │ Any      │
├─────┼──────────┼──────────┼──────────┼──────────┤
│ 1   │ b        │ 0.633028 │ 0.388354 │ 0.134275 │
│ 2   │ c        │ foo      │ 2.2      │ 1        │

julia> df_floats = DataFrame(a=[:x, :y, :z], b=rand(3), c=rand(3))
3×3 DataFrame
│ Row │ a      │ b        │ c        │
│     │ Symbol │ Float64  │ Float64  │
├─────┼────────┼──────────┼──────────┤
│ 1   │ x      │ 0.941859 │ 0.35769  │
│ 2   │ y      │ 0.404513 │ 0.237959 │
│ 3   │ z      │ 0.649489 │ 0.255324 │

julia> DataFrames.transpose1(df_floats)
2×4 DataFrame
│ Row │ variable │ x        │ y        │ z        │
│     │ String   │ Float64? │ Float64? │ Float64? │
├─────┼──────────┼──────────┼──────────┼──────────┤
│ 1   │ b        │ 0.941859 │ 0.404513 │ 0.649489 │
│ 2   │ c        │ 0.35769  │ 0.237959 │ 0.255324 │

julia> DataFrames.transpose2(df_floats)
2×4 DataFrame
│ Row │ variable │ x        │ y        │ z        │
│     │ String   │ Float64  │ Float64  │ Float64  │
├─────┼──────────┼──────────┼──────────┼──────────┤
│ 1   │ b        │ 0.941859 │ 0.404513 │ 0.649489 │
│ 2   │ c        │ 0.35769  │ 0.237959 │ 0.255324 │

julia> DataFrames.transpose3(df_floats)
2×4 DataFrame
│ Row │ variable │ x        │ y        │ z        │
│     │ String   │ Float64  │ Float64  │ Float64  │
├─────┼──────────┼──────────┼──────────┼──────────┤
│ 1   │ b        │ 0.941859 │ 0.404513 │ 0.649489 │
│ 2   │ c        │ 0.35769  │ 0.237959 │ 0.255324 │

julia> df_number = DataFrame(a=[:x, :y, :z], b=rand(3), c=rand(Int, 3))
3×3 DataFrame
│ Row │ a      │ b         │ c                   │
│     │ Symbol │ Float64   │ Int64               │
├─────┼────────┼───────────┼─────────────────────┤
│ 1   │ x      │ 0.438549  │ 480681029893134526  │
│ 2   │ y      │ 0.489623  │ 172366260093707646  │
│ 3   │ z      │ 0.0132915 │ 3468722734556270181 │

julia> DataFrames.transpose1(df_number)
2×4 DataFrame
│ Row │ variable │ x          │ y          │ z          │
│     │ String   │ Float64?   │ Float64?   │ Float64?   │
├─────┼──────────┼────────────┼────────────┼────────────┤
│ 1   │ b        │ 0.438549   │ 0.489623   │ 0.0132915  │
│ 2   │ c        │ 4.80681e17 │ 1.72366e17 │ 3.46872e18 │

julia> DataFrames.transpose2(df_number)
2×4 DataFrame
│ Row │ variable │ x          │ y          │ z          │
│     │ String   │ Float64    │ Float64    │ Float64    │
├─────┼──────────┼────────────┼────────────┼────────────┤
│ 1   │ b        │ 0.438549   │ 0.489623   │ 0.0132915  │
│ 2   │ c        │ 4.80681e17 │ 1.72366e17 │ 3.46872e18 │

julia> DataFrames.transpose3(df_number)
2×4 DataFrame
│ Row │ variable │ x          │ y          │ z          │
│     │ String   │ Float64    │ Float64    │ Float64    │
├─────┼──────────┼────────────┼────────────┼────────────┤
│ 1   │ b        │ 0.438549   │ 0.489623   │ 0.0132915  │
│ 2   │ c        │ 4.80681e17 │ 1.72366e17 │ 3.46872e18 │
```

Still lots to do here, but I thought it would be worth it to get the ball rolling and see if people have any thoughts. I will probably get back to this in earnest next week.

bkamins commented:
Thank you for working on it. To get a safe eltype of the resulting columns using `typejoin` should be a safe choice I think.

The issues you raise are exactly the things that I expected/feared (i.e. that it would be hard to find the optimal solution)

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,20 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+transpose1(df::AbstractDataFrame, indexcol=1) = unstack(stack(df,Not(indexcol)), :variable, indexcol,:value)
+
+function transpose2(df::AbstractDataFrame, indexcol=1)
+    m = permutedims(Matrix(df[!, Not(indexcol)]))
```
here you can use `typejoin` to find an appropriate type.

nalimilan replied:
`promote` is probably even better as it will take care of choosing the best numeric types if numeric columns have different types. It will also return `Union{T,Missing}` rather than `Any`.

nalimilan replied:
Actually `Matrix(::DataFrame)` already uses `promote_type`. AFAICT the behavior is correct. You get an `Any` column only because one of the input columns is `Anÿ`, but that's expected.

bkamins replied:
I proposed `typejoin` so that taking transpose of a transpose does not change values and gets back where we started. But maybe it is OK to make a promotion (for a `Matrix` for sure it makes sense, but I was not sure here)

nalimilan replied:
Ah yes. But getting concrete types sounds more important?

bkamins replied:
Let us discuss it, as probably most important problematic case is `Bool` (as you later might want to use it to check logical conditions):
```
julia> df = DataFrame([[true,false],1:2,1.0:2.0])
2×3 DataFrame
│ Row │ x1   │ x2    │ x3      │
│     │ Bool │ Int64 │ Float64 │
├─────┼──────┼───────┼─────────┤
│ 1   │ 1    │ 1     │ 1.0     │
│ 2   │ 0    │ 2     │ 2.0     │

julia> Matrix(df)
2×3 Array{Float64,2}:
 1.0  1.0  1.0
 0.0  2.0  2.0
```

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,20 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
 end
+
+
+transpose1(df::AbstractDataFrame, indexcol=1) = unstack(stack(df,Not(indexcol)), :variable, indexcol,:value)
```
in general - the output index column name should be passed independently + some check if it conflicts with the generated column names should be made.

kescobo replied:
Agree - was thinking it might be worth putting the `makeunique` keyword in there to deal with cases where the index column is not all unique.

bkamins replied:
yes, this is a standard we do 😄.

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,20 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+function transpose2(df::AbstractDataFrame, indexcol=1)
+    m = permutedims(Matrix(df[!, Not(indexcol)]))
+    df2 = DataFrame(variable=names(df[!, Not(indexcol)]))
+    hcat(df2, DataFrame(m, df[!, indexcol]))
```
`copycols=false` would help performance a bit here.

Above, we could also try to avoid doing two copies (one due to `Matrix` and the other due to `permutedims`), but that doesn't seem super easy if we still want to take advantage of optimized `permutedims` methods.

kescobo commented:
Aright, doing a bit more on this. 

1. Changed function signature and included destination col argument
2. Added some other signatures to try out. I put some benchmark code in a [gist here](https://gist.github.com/kescobo/b248b529e453e207b8bea0d43f6fd25d)

Results are for `Any` type (a), numerical types including Bool (n), and for dataframes where rows have mixed type (some all numerical, some with strings) for different numbers of rows.

This is basically the same as before.

```
┌ Warning: a10
└ @ Main ~/repos/DataFramesDev/benchmarks.jl:17
  35.006 μs (401 allocations: 32.67 KiB)
  11.873 μs (160 allocations: 16.30 KiB)
  14.772 μs (166 allocations: 16.39 KiB)
┌ Warning: a100
└ @ Main ~/repos/DataFramesDev/benchmarks.jl:21
  136.785 μs (1086 allocations: 115.98 KiB)
  46.247 μs (541 allocations: 58.39 KiB)
  143.145 μs (1255 allocations: 118.94 KiB)
┌ Warning: a1k
└ @ Main ~/repos/DataFramesDev/benchmarks.jl:25
  1.157 ms (7927 allocations: 976.36 KiB)
  428.862 μs (4669 allocations: 533.55 KiB)
  1.529 ms (20029 allocations: 1.27 MiB)
┌ Warning: n10
└ @ Main ~/repos/DataFramesDev/benchmarks.jl:29
  34.351 μs (361 allocations: 32.05 KiB)
  12.244 μs (140 allocations: 15.98 KiB)
  15.950 μs (166 allocations: 16.39 KiB)
┌ Warning: n100
└ @ Main ~/repos/DataFramesDev/benchmarks.jl:33
  132.626 μs (686 allocations: 109.73 KiB)
  45.631 μs (341 allocations: 55.27 KiB)
  151.399 μs (1255 allocations: 118.94 KiB)
┌ Warning: n1k
└ @ Main ~/repos/DataFramesDev/benchmarks.jl:37
  1.076 ms (3927 allocations: 913.86 KiB)
  429.272 μs (2669 allocations: 502.30 KiB)
  1.633 ms (20029 allocations: 1.27 MiB)
┌ Warning: mr10
└ @ Main ~/repos/DataFramesDev/benchmarks.jl:41
  33.293 μs (407 allocations: 32.77 KiB)
  12.441 μs (160 allocations: 16.30 KiB)
  15.000 μs (166 allocations: 16.39 KiB)
┌ Warning: mr100
└ @ Main ~/repos/DataFramesDev/benchmarks.jl:45
  135.343 μs (1121 allocations: 116.53 KiB)
  46.999 μs (541 allocations: 58.39 KiB)
  143.275 μs (1255 allocations: 118.94 KiB)
┌ Warning: mr1k
└ @ Main ~/repos/DataFramesDev/benchmarks.jl:49
  1.191 ms (8261 allocations: 981.58 KiB)
  438.124 μs (4669 allocations: 533.55 KiB)
  1.546 ms (20029 allocations: 1.27 MiB)
```

A couple of new observations:

1. with @nalimilan 's suggestion for `copycols` for the permuted matrix, there's a moderate speed-up. Though I had to remove the inner  `copycols` argument (for the constructor with the matrix), as it was causing a method error

```
julia> @btime DataFrames.transpose2($mr1k, copycols=false);
  425.934 μs (4669 allocations: 533.55 KiB)

julia> @btime DataFrames.transpose2($mr1k, copycols=true);
  459.766 μs (5672 allocations: 643.16 KiB)
```
With both copycols:
```
ERROR: MethodError: no method matching DataFrame(::Matrix{Any}, ::Vector{String}; copycols=true)
Closest candidates are:
  DataFrame(::AbstractMatrix{T} where T, ::AbstractVector{var"#s25"} where var"#s25"<:AbstractString; makeunique) at /home/kevin/repos/DataFramesDev/dev/DataFrames.jl/src/dataframe/dataframe.jl:272 got unsupported keyword argument "copycols"
```

2. I made a comprehension version of both `transpose2` and `transpose3` to help with types for mixed row columns. Not surprisingly, this slows down t2 (though it's still faster than t3), but I was surprised that it actually speeds up t3. That is, doing `Vector(row)` is slower than `[x for x in row]`

```
julia> @btime DataFrames.transpose2_comprehension($mr100);
  113.745 μs (1659 allocations: 116.34 KiB)

julia> @btime DataFrames.transpose2($mr100);
  48.217 μs (541 allocations: 58.39 KiB)

julia> @btime DataFrames.transpose3($mr100);
  144.665 μs (1255 allocations: 118.94 KiB)

julia> @btime DataFrames.transpose3_comprehension($mr100);
  125.924 μs (2356 allocations: 168.48 KiB)
```

Both give columns of `Real` when the row has integers and floats, and `Any` when there are `String`s in the row.

I'm not totally clear on where to add the `promote` or `typejoin` calls for `transpose2`. I tried a couple of places, and it doesn't seem to alter the output type of the columns. The promotion already seems to happen with the Matrix conversion, and can't be fixed if there are Strings in the matrix:

```
julia> n10
10×4 DataFrame
│ Row │ a      │ b         │ c                    │ d    │
│     │ String │ Float64   │ Int64                │ Bool │
├─────┼────────┼───────────┼──────────────────────┼──────┤
│ 1   │ a17Wik │ 0.356758  │ -8986863187679119824 │ 0    │
│ 2   │ os6kQD │ 0.354969  │ -1901978611141674116 │ 0    │
│ 3   │ PoiIab │ 0.769343  │ 4759929156393762244  │ 1    │
│ 4   │ 2XDnn1 │ 0.0322987 │ 2546570589132666494  │ 0    │
│ 5   │ ZIx6QI │ 0.560122  │ 6388359009763183408  │ 1    │
│ 6   │ MO1iLD │ 0.194934  │ -1409080130674345575 │ 0    │
│ 7   │ zrhAak │ 0.981582  │ -2685183505188918570 │ 0    │
│ 8   │ RZtUU1 │ 0.61182   │ -3456256066328160472 │ 1    │
│ 9   │ eoMStL │ 0.024695  │ 8468153461503185050  │ 1    │
│ 10  │ rar8je │ 0.324927  │ -983327689695525853  │ 1    │

julia> eltype(Matrix(n10[!, Not(:a)]))
Float64
```
I had some ideas with `promote`, but it throws if there are any strings encountered, and checking for mismatched types seems like it would eat any performance benefits of the permuted matrix approach.

```
julia> promote([1, 1., "str"]...)
ERROR: promotion of types Int64, Float64 and String failed to change any arguments
```

Very possible I'm missing something.

bkamins commented:
> The promotion already seems to happen with the Matrix conversion

Yes. To avoid promotion you would have to write something like `Matrix{Any}(...)`.

So the conclusion is that `Matrix` is fastest, right? Then I would go for it and add an additional kwarg `promote_type::Bool`. If yes - then call `Matrix` (and then comprehension is not needed), if `false` then call `Matrix{Any} (and use comprehension). Just please make sure to pass `Tables.table` wrapped matrix to the constructor (cf. #2464)

In general an input from @nalimilan would be welcome on naming kwargs in this function (he is far superior than me in such things).

kescobo commented:
> So the conclusion is that Matrix is fastest, right?

Yes, by a fair amount.

> Then I would go for it and add an additional kwarg promote_type::Bool. If yes - then call Matrix (and then comprehension is not needed), if false then call Matrix{Any} (and use comprehension).

This makes sense to me. Do you have a sense of whether the default for promotion should be `true` or `false`?

bkamins commented:
> Do you have a sense of whether the default for promotion should be `true` or `false`?

Not sure - @nalimilan seemed to prefer `true`, but for me intuitively `false` would be safer.

Having said that I would go for `true` as I assume that most of the time when you transpose the data frame holds homogenous data in the transformed columns.

kescobo commented:
Open questions

1. name and default for `promote_type` kwarg
2. `src_colname` and `dest_colname` should be positional, right?
3. I'm also wondering if `dest_colname` is necessary. I'd be in favor of having this be the name of the column that gets indexed on, and have users need to explicitly rename
4. Should `copycols` always be `false`? The matrix gets allocated either way, and so there's no safety issue, right?

I wrote some docs and tests, but will wait to finalize until the above are answered. Any additional feedback appreciated

kescobo commented:
I'm getting some weird segfaults trying to test locally that don't appear to be happening on CI. Is it possible that Revise is causing issues?

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,40 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+
+"""
+    transpose(df::AbstractDataFrame, src_namescol=1, dest_namescol=names(df)[src_namescol]; copycols::Bool=false, makeunique=false, promote_type::Bool=true)
```
please make sure to obey 80 characters per line as a soft limit and 92 as a hard limit

bkamins commented on src/abstractdataframe/reshape.jl:
`````diff
@@ -399,3 +399,40 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+transpose(df2, promote_type=false)
+````
+"""
+function Base.transpose(df::AbstractDataFrame, src_namescol=1, dest_namescol=names(df)[src_namescol]; copycols::Bool=false, makeunique=false, promote_type::Bool=true)
`````
having `src_namescol` and `dest_namescol` positional is OK I think. I am not 100% sure about the defaults (and if we should provide them, especially for `dest_namescol`), but let us keep what you propose. `copycols` is not needed. Here is a proposal of a signature:
```suggestion
function Base.transpose(df::AbstractDataFrame, src_namescol::ColumnIndex=1,
                        dest_namescol::Union{Symbol, AbstractString}=src_namescol isa Integer ?
                                                                     _names(df)[src_namescol] :
                                                                     src_namescol;
                        makeunique::Bool=false, promote_type::Bool=true)
```

bkamins commented on src/abstractdataframe/reshape.jl:
`````diff
@@ -399,3 +399,40 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+````
+"""
+function Base.transpose(df::AbstractDataFrame, src_namescol=1, dest_namescol=names(df)[src_namescol]; copycols::Bool=false, makeunique=false, promote_type::Bool=true)
+    if promote_type
`````
maybe create `df[!, Not(src_namescol)]` once above and reuse it?

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,40 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+"""
+    transpose(df::AbstractDataFrame, src_namescol=1, dest_namescol=names(df)[src_namescol]; copycols::Bool=false, makeunique=false, promote_type::Bool=true)
+
+Transpose a `DataFrame`, such that rows become columns,
```
technically what we do is `permutedims`, but probably it is better to call it `transpose` as you propose. Just please make a note in the docstring that `transpose` is not recursive (as opposed to `transpose` in Base).

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,40 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    if promote_type
+        m = permutedims(Matrix(df[!, Not(src_namescol)]))
+        df2 = DataFrame([names(df[!, Not(src_namescol)])], [dest_namescol])
+        return hcat(df2, DataFrame(m, df[!, src_namescol], makeunique=makeunique), copycols=copycols)
```
better use `hcat!`, `copycols` can be `false` here. Also note that `hcat!` also has `makeunique` kwarg that should be used.

Also - as commented `DataFrame(m, df[!, src_namescol], makeunique=makeunique)` will soon be deprecated, so we should write something like `df_tmp = rename!(DataFrame(Tables.table(m)), df[!, src_namescol]))` and then use it in `hcat!`.

Finally the decision is if we require `df[!, src_namescol]` to hold `Symbol` or `AbstractString` (I think yes) or we should do a conversion to `Symbol` of whatever is stored in this column.

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,40 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+        df2 = DataFrame([names(df[!, Not(src_namescol)])], [dest_namescol])
+        return hcat(df2, DataFrame(m, df[!, src_namescol], makeunique=makeunique), copycols=copycols)
+    else
+        m = permutedims(Matrix{Any}(df[!, Not(src_namescol)]))
```
similar comments as above

bkamins commented on test/reshape.jl:
```diff
@@ -507,4 +507,45 @@ end
+@testset "transpose" begin
+    df1 = DataFrame(a=["x", "y"], b=rand(2), c=[1,2], d=rand(Bool,2))
+    df2 = DataFrame(a=["x", "y"], b=[1., "str"], c=[1,2], d=rand(Bool,2))
+    df3 = DataFrame(a=fill("x", 10), b=rand(10), c=rand(Int, 10), d=rand(Bool,10))
```
please check:
1) `DataFrame()`
2) `DataFrame(a=Int[], b=Float64[])`
3) and cases that should error (e.g. missing column, column with new names that is not `Symbol` or string etc.)

kescobo replied:
Just to be sure, when would expect:

1. `permutedims(DataFrame()) == DataFrame()` or error?
2. I'm thinking error, since there are no values for headers?
3. when you say "missing column", do you mean where the `src_namescol` has a `missing`? Or where certain columns have `missing` values?

bkamins replied:
Ad.1 - it should error I think
Ad 3. - when I say "missing column" I mean "when the requested column is not present in a data frame"

bkamins commented:
>  I'd be in favor of having this be the name of the column that gets indexed on, and have users need to explicitly rename.

It is necessary as the name of `src_colname` might conflict with the column names automatically generated. Assume `src_colname` has name `:a` and contains a vector `[:a, :b, :c]` for instance.

> Is it possible that Revise is causing issues?

Not sure - it should not be an issue.

kescobo commented:
Great points, thanks for review! I will try to address early next week 👍

nalimilan commented:
Note that in theory `promote_type::Bool` isn't enough to cover all cases, as several behaviors are possible:
- choose the element type by calling `promote_type` on the column's eltypes
- choose the element type by calling `typejoin` on the column's eltypes
- choose the element type by calling `collect` or a comprehension on the columns to take into account the type of the values they actually contain (this calls `Base.promote_typejoin` under the hood; we don't have a way to do the same thing using `promote_type` at the moment)

I'm not sure how useful the second option could be in practice, but we could consider supporting it too, for example by allowing to pass any function to compute the type to an argument called `eltype`. Not sure how the last behavior would be specified, but many options are possible (symbol, `nothing`, `missing`, `collect`...).

bkamins commented:
> Not sure how the last behavior would be specified, but many options are possible

So maybe let us make a `promotion::Symbol` kwarg, now implement what we feel we know we want, and this will allow to add more options later?

bkamins commented on src/abstractdataframe/reshape.jl:
`````diff
@@ -399,3 +399,40 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+````
+"""
+function Base.transpose(df::AbstractDataFrame, src_namescol=1, dest_namescol=names(df)[src_namescol]; copycols::Bool=false, makeunique=false, promote_type::Bool=true)
+    if promote_type
`````
yes - it should be relatively cheap to check in run-time

kescobo replied:
> Note that if all columns have a concrete element type

to clarify, you mean if they all have _the same_ concrete eltype?

bkamins commented:
> I'd be more included to overload `permutedims`

This is what I thought you would prefer (that is why I have raised it). Let us then switch to `permutedims` (though it is a pity that `transform` cannot be used).

kescobo commented:
> Let us then switch to permutedims (though it is a pity that transform cannot be used).

Agree it's a pity, but makes sense. Main downside from my perspective is discoverability - what do we think about defining something like:

```
Base.transpose(::AbstractDataFrame, args...; kwargs...) = MethodError("transpose not defined for AbstractDataFrames. Try permutedims instead")
```

bkamins commented:
I think it is a good idea.

kescobo commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,40 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    else
+        m = permutedims(Matrix{Any}(df[!, Not(src_namescol)]))
+        df2 = DataFrame([names(df[!, Not(src_namescol)])], [dest_namescol])
+        hcat(df2, DataFrame([[x for x in col] for col in eachcol(m)], df[!, src_namescol], makeunique=makeunique), copycols=copycols)
```
Evidently not:

```
julia> m = Any[missing 1.; 2. 3.]
2×2 Matrix{Any}:
  missing  1.0
 2.0       3.0

julia> collect.(eachcol(m))
2-element Vector{Vector{Any}}:
 [missing, 2.0]
 [1.0, 3.0]

julia> [[x for x in col] for col in eachcol(m)]
2-element Vector{Vector{T} where T}:
 Union{Missing, Float64}[missing, 2.0]
 [1.0, 3.0]
```

nalimilan replied:
Ah right, `collect` preserves the eltype when it's defined. Kind of inconsistent...

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -433,14 +433,16 @@ function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex=1,
                                                  src_namescol;
     makeunique::Bool=false, promote::Symbol=:all)
 
+    nrow(df) > 0 || throw(ArgumentError("`permutedims` not defined for tables with 0 rows"))
```
you can use `1` as a default, but in the body of the function just do the check:
```
int_col = src_namescol isa Integer ? src_namescol : columnindex(src_namescol)
1 <= int_col <= ncol(df) || throw(ArgumentError(...))
```

kescobo replied:
Sorry, I should clarify - the `BoundsError` is coming from the default for `dest_namescol`:

```
dest_namescol::Union{Symbol, AbstractString}=src_namescol isa Integer ?
                                                 _names(df)[src_namescol] :
                                                 src_namescol
```

bkamins replied:
ah - then maybe simply define a separate method without `dest_namescol` argument that would do proper error handling and dispatch to the main method on success?

kescobo replied:
Ah, right. Good point.

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,51 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+Transpose a `DataFrame`, such that rows become columns,
+and the column indexed by `src_namescol` becomes a header.
+
+By default, the type of resulting columns will depend on the promoted type of all input columns.
+Pass `promote_type=false` to make resulting column types depend on the elements of the originating rows,
+though note that this may be substantially slower.
```
Actually I wonder whether we really need to add this argument now. Could you leave it out unless we have a reason to think the default isn't good enough? In general data frames should have concretely-types columns, and in that case promotion will give the best result.

kescobo replied:
I think the reason for argument is, if columns being permuted have mixed types, you may have to chose between fast and concrete type:

```
julia> df = DataFrame(a=[randstring(5) for _ in 1:100], b=rand(100), c=repeat([:x, 1.], inner=50));

julia> pd_all = @btime permutedims(df, promote=:all);
  41.602 μs (410 allocations: 50.12 KiB)

julia> pd_none = @btime permutedims(df, promote=:none);
  89.565 μs (1070 allocations: 84.80 KiB)

julia> eltype.(eachcol(pd_none))
101-element Vector{DataType}:
 String
 Any
 Any
 Any
 Any
 Any
 ⋮
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64

julia> eltype.(eachcol(pd_all))
101-element Vector{DataType}:
 String
 Any
 Any
 Any
 Any
 Any
 ⋮
 Any
 Any
 Any
 Any
 Any
 Any
```

nalimilan replied:
Yes but is that really common? If you permute your data frame, it sounds likely that the data is homogeneous.

bkamins replied:
@kescobo - what @nalimilan means is that we can now implement only the default behaviour (leaving out the kwarg). The PR then can be merged, and later we can open another issue/PR to discuss how to best handle other options as this will be non-breaking.

kescobo replied:
> Yes but is that really common?

@nalimilan For me, it's pretty common. There are some tools in my field that expect stuff in column-major format, and others that expect column-major. I often write my own version of this function to convert between the two.

But it's easy enough to concretize after the fact, so I'm fine with this.

kescobo commented on test/reshape.jl:
```diff
@@ -507,45 +507,63 @@ end
+    @test size(df1_pd, 1) == ncol(df1) - 1
+    @test size(df1_pd, 2) == nrow(df1) + 1
+    @test names(df1_pd) == ["a", "x", "y"]
+    @test eltype(df1_pd.x) <: AbstractFloat
```
The reason I did it this way instead of `Float64` is that, I assumed on 32 bit machines, this would be `Float32`. Is there another way to deal with that, or we just don't test on 32 bit?

nalimilan replied:
No, `Float64` is also used in 32-bit machines.

kescobo replied:
TIL!

kescobo commented:
Alright, I _think_ I addressed everything brought up so far. Apologies if I resolved comments that weren't addressed.

I've written a docstring (and converted example to doctest per @nalimilan 's suggestion), but should I also add a section to the main docs (assuming in "Reshaping and pivoting").

kescobo commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+Base.transpose(::AbstractDataFrame, args...; kwargs...) =
+    MethodError("`transpose` not defined for `AbstractDataFrame`s. Try `permutedims` instead")
+    # ↑ is 94 char, is that OK? (there are a couple other lines in this file that go over)
```
Oh - and I'm not 100% clear on the character limit... there are a couple of places in this file that go over - was that comment just about the docstring?

bkamins replied:
Here are the style guidelines: https://github.com/JuliaData/DataFrames.jl/blob/master/CONTRIBUTING.md#style-guidelines

(so in general it should be everywhere, but in practice we are not 100% consistent as sometimes keeping narrow lines looks ugly and this package is quite old and had many contributors :smile:)

bkamins replied:
94 is OK I think in this case

bkamins commented:
> assuming in "Reshaping and pivoting"

sure - it would be nice to have a short entry there.

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+"""
+    permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex,
+                dest_namescol::Union{Symbol,Int};
```
```suggestion
                dest_namescol::Union{Symbol,AbstractString};
```

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex; makeunique::Bool=false)
+    permutedims(df::AbstractDataFrame; makeunique::Bool=false)
+
+Turn a `DataFrame` on its side such that rows become columns
```
(it can be also `SubDataFrame`)
```suggestion
Turn `df` on its side such that rows become columns
```

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    permutedims(df::AbstractDataFrame; makeunique::Bool=false)
+
+Turn a `DataFrame` on its side such that rows become columns
+and the column indexed by `src_namescol` becomes a header.
```
maybe say not "header" but "column names" (we do not use "header" anywhere).
But treat my language corrections lightly, as @nalimilan is much better than me in such things.

Also mention that `src_namescol` will be removed from a data frame.

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+  with `_i` (`i` starting at 1 for the first duplicate).
+
+Note: The eltypes of columns in resulting `DataFrame` will depend
+on the eltypes of _all_ input columns.
```
except `src_namescol`

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+Note: The eltypes of columns in resulting `DataFrame` will depend
+on the eltypes of _all_ input columns.
+That is, if the source `DataFrame` contains `String` and `Int` columns,
```
maybe mention that `promote_type` is used?

bkamins commented on src/abstractdataframe/reshape.jl:
````diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+# Examples
+
+```jldoctest
+julia> df1 = DataFrame(a=["x", "y"], b=rand(2), c=[1,2], d=rand(Bool,2));
````
do not use random values (I know we used them in the past in some places, but it is not reproducible).
Also I think showing the source data frames is in this case desirable to visualize the difference.

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+# Arguments
+- `df` : the `AbstractDataFrame`
+- `src_namescol` : the column that will become the new header.
```
add information what are the requirements on values stored in this column (or that there are no requiremens)

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+                          dest_namescol::Union{Symbol, AbstractString};
+                          makeunique::Bool=false)
+
+    nrow(df) > 0 || throw(ArgumentError("`permutedims` not defined for data frame with 0 rows"))
```
why is this not allowed? Technically you can permute it and produce a data frame with  one column: `dest_namecol`

kescobo replied:
:thinking: I thought we said above that it would error... Perhaps I misunderstood.

bkamins replied:
You asked above about `DataFrame()` which has `ncol(df)==0` (not `nrow`). Let us just decide what is best. I commented because it seems that `nrow(df)==0` if `ncol(df)>0` is not a problem (not very useful though - it will be just a column with old column names - right?)

kescobo replied:
Makes sense, I added support for this (also needed fix for single column) - more discussion below :point_down:

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+    nrow(df) > 0 || throw(ArgumentError("`permutedims` not defined for data frame with 0 rows"))
+    eltype(df[!, src_namescol]) <: SymbolOrString || throw(
+            ArgumentError("src_namescol must have eltype `Symbol` or `AbstractString`"))
```
```suggestion
            ArgumentError("src_namescol must have eltype `Symbol` or `<:AbstractString`"))
```

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+            ArgumentError("src_namescol must have eltype `Symbol` or `AbstractString`"))
+
+    df_notsrc = df[!, Not(src_namescol)]
+    df_permuted = DataFrame([names(df_notsrc)], [dest_namescol])
```
this seems cleaner:
```suggestion
    df_permuted = DataFrame(dest_namescol => names(df_notsrc))
```

bkamins replied:
also a decision to be made is if we want to produce `String`s or `Symbol`s here (see a similar situation in `stack`/`unstack` - maybe we should have a kwarg allowing to decide it - and then what should be the default?).

kescobo replied:
Will leave this open waiting for @nalimilan to respond. My take is that since `names` went to strings, we should use strings. I'm happy to add a keyword though if there's consensus (should maybe allow `Int` as a possibility too in that case?)

bkamins replied:
Sure. In `stack` we default to `String`, so for now we could leave it as is and add a kwarg later if needed (similarly to promotion).

nalimilan replied:
Agreed. Contrary to `stack`, each name appears only once, so there's little point in using a `PooledArray` or a `CategoricalArray`.

bkamins replied:
But the point is not about being pooled, but if the column should be of eltype: `String` or `Symbol` (in `stack` we allow the user to choose one of them).

nalimilan replied:
Yes, I know, I was commenting on the fact that a keyword argument is less needed because `CategoricalValue` is less useful. Indeed we could add an argument later to get symbols instead of strings, but that doesn't sound too important.

kescobo replied:
:+1: OK, will leave as-is for now.

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+    m = permutedims(Matrix(df_notsrc))
+    df_tmp = rename!(DataFrame(Tables.table(m)), df[!, src_namescol], makeunique=makeunique)
+    hcat!(df_permuted, df_tmp, copycols=false)
```
```suggestion
    return hcat!(df_permuted, df_tmp, copycols=false)
```

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex;
+                          makeunique::Bool=false)
+    nrow(df) > 0 || throw(ArgumentError("`permutedims` not defined for data frame with 0 rows"))
```
no need to check it - it will be checked in the other function anyway

kescobo replied:
Ah, I was confusing it with the zero column case :+1:

bkamins replied:
just commented that above :smile:

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    permutedims(df, src_namescol, dest_namescol; makeunique=makeunique)
+end
+
+function Base.permutedims(df::AbstractDataFrame; makeunique::Bool=false)
+    permutedims(df, 1; makeunique=makeunique)
+end
```
we typically use short form in such cases:
```suggestion
Base.permutedims(df::AbstractDataFrame; makeunique::Bool=false) =
    permutedims(df, 1; makeunique=makeunique)
```

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    else
+        dest_namescol = src_namescol
+    end
+    permutedims(df, src_namescol, dest_namescol; makeunique=makeunique)
```
```suggestion
    return permutedims(df, src_namescol, dest_namescol; makeunique=makeunique)
```

bkamins commented:
@kescobo - please push things you accept in a batch (as otherwise CI will get swamped by many small changes for many hours)

bkamins commented on test/reshape.jl:
```diff
@@ -507,4 +507,58 @@ end
+    @test_throws ArgumentError permutedims(df4[!, [:a,:b,:c]])
+    # Can't index in the presence of missing
+    @test_throws ArgumentError permutedims(df4[!, [:g, :a, :b, :c]])
+    @test_throws ArgumentError permutedims(df4[!, [:h,:a,:b]])
```
can you please make tests consistent to use space after `,` everywhere (again - I know we do not obey this rule, but the places that do not follow it should be cleaned up)

kescobo replied:
Well, I cleaned it up in this file... regex replace is a wonderful thing :laughing:

bkamins commented on test/reshape.jl:
```diff
@@ -507,4 +507,58 @@ end
 end
 
+@testset "permutedims" begin
+    df1 = DataFrame(a=["x", "y"], b=rand(2), c=[1,2], d=rand(Bool,2))
```
it would be easier to read tests if you followed the data frame definition immediately by the tests using it.

kescobo commented:
> @kescobo - please push things you accept in a batch (as otherwise CI will get swamped by many small changes for many hours)

Yeah, sorry, meant to accept your suggestions in batch, didn't realize I had to go through file view for that.

kescobo commented:
Aright, packing it in for the day. Will summarize the stuff that's remaining in the morning.

bkamins commented:
> meant to accept your suggestions in batch

Sure - great. I stress this so much, because I guess DataFrames.jl is the second after Julia Base in terms of the amount of tests we run.

bkamins commented:
Also please add a reference to `permutedims` in functions.md, as otherwise docs will not build.

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,109 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+- `src_namescol` : the column that will become the new header.
+  This column eltype must be `<: Union{String, Symbol}`.
+  Defaults to first column.
+- `dest_namescol` : the name of the first column in the returned `DataFrame`.
+  Defaults to the same name as `src_namescol`.
```
Does this default makes sense in practice? Do we have examples of similar functions in other languages?

bkamins replied:
I do not think we have alternative examples, but it kind of makes sense because such a column is most naturally called `"row_names"` or something like this and after transpose it is still row names (complementing column names).

kescobo replied:
Agree. An alternative way of thinking is that if you're looking at a table of variables that are stored in columns, and you need them stored in rows (my most typical use), the first column name becomes the first row name. Eg

| sample | obs1 | obs2 |
|-- |--|--|
| a | 1.2 | 3.14 |
| b | 42.0 | 1.0 |

becomes

| sample | a | b |
|--|--|--|
| obs1 | 1.2 | 42.0|
| obs2 | 3.14 | 1.0|

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,109 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+  if duplicate names are found; if `true`, duplicate names will be suffixed
+  with `_i` (`i` starting at 1 for the first duplicate).
+
+Note: The eltypes of columns in resulting `DataFrame`
+(other than the first column, which always has eltype `String`)
+will depend on the eltypes of _all_ input columns
+based on the results of `prote_type`.
```
```suggestion
Note: The element types of columns in resulting `DataFrame`
(other than the first column, which always has element type `String`)
will depend on the element types of _all_ input columns
based on the result of `promote_type`.
```

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,109 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+(other than the first column, which always has eltype `String`)
+will depend on the eltypes of _all_ input columns
+based on the results of `prote_type`.
+That is, if the source `DataFrame` contains `String` and `Int` columns,
+resulting columns will have eltype `Any`.
+If the source has a mix of numeric types (eg. `Float64` and `Int`),
+columns in resulting `DataFrame` will be promoted to `Float64`.
```
```suggestion
That is, if the source data frame contains `Int` and `Float64` columns,
resulting columns will have element type `Float64`. If the source has
`Int` and `String` columns, resulting columns will have element type `Any`.
```
(some numeric types promote to other types, e.g. to `BigFloat`)

kescobo replied:
:+1:

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,109 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+                          dest_namescol::Union{Symbol, AbstractString};
+                          makeunique::Bool=false)
+
+    nrow(df) > 0 || throw(
+        ArgumentError("`permutedims` not defined for data frame with 0 rows"))
+    eltype(df[!, src_namescol]) <: SymbolOrString || throw(
+            ArgumentError("src_namescol must have eltype `Symbol` or `<:AbstractString`"))
```
```suggestion
    nrow(df) > 0 ||
        throw(ArgumentError("`permutedims` not defined for data frame with 0 rows"))
    eltype(df[!, src_namescol]) <: SymbolOrString ||
        throw(ArgumentError("src_namescol must have eltype `Symbol` or `<:AbstractString`"))
```

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,109 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    return permutedims(df, src_namescol, dest_namescol; makeunique=makeunique)
+end
+
+Base.permutedims(df::AbstractDataFrame; makeunique::Bool=false) =
+    permutedims(df, 1; makeunique=makeunique)
```
Why not just use `src_namescol::ColumnIndex=1` in the last definition?

kescobo replied:
I swear there was a reason... either it was wrong or doesn't apply anymore

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,109 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex;
+                          makeunique::Bool=false)
+    if src_namescol isa Integer
+        1 <= src_namescol <= ncol(df) || throw(ArgumentError("`src_namescol` doesn't exist"))
```
I think `BoundsError` like `getindex` throws would be more informative.

nalimilan commented on test/reshape.jl:
```diff
@@ -507,4 +507,62 @@ end
+    for (i, row) in enumerate(eachrow(df2_pd))
+        @test Vector(row) == [orignames2[i]; df2[!, orignames2[i]]]
+    end
+    @test Any <: eltype(df2_pd.x)
+    @test Any <: eltype(df2_pd.y)
```
```suggestion
    @test eltype(df2_pd.x) == Any
    @test eltype(df2_pd.y) == Any
```

nalimilan commented on test/reshape.jl:
```diff
@@ -507,4 +507,62 @@ end
+    @test permutedims(df4[!, [:a, :b, :c, :e]], :e) ==
+          permutedims(df4[!, [:e, :a, :b, :c]]) ==
+          permutedims(df4[!, [:a, :b, :c, :f]], :f, :e)
+    # Can't index Float Column
```
```suggestion
    # Can't index float column
```

nalimilan commented on test/reshape.jl:
```diff
@@ -507,4 +507,62 @@ end
+        @test Vector(row) == [orignames1[i]; df1[!, orignames1[i]]]
+    end
+
+    @test eltype(df1_pd.x) <: Float64
+    @test eltype(df1_pd.y) <: Float64
```
```suggestion
    @test eltype(df1_pd.x) == Float64
    @test eltype(df1_pd.y) == Float64
```

nalimilan commented on test/reshape.jl:
```diff
@@ -507,4 +507,62 @@ end
+          permutedims(df4[!, [:a, :b, :c, :f]], :f, :e)
+    # Can't index Float Column
+    @test_throws ArgumentError permutedims(df4[!, [:a, :b, :c]])
+    # Can't index in the presence of missing
```
```suggestion
    # Can't index columns that allow for missing
```
(even if there aren't any missing values; we may want to relax this in the future)

kescobo commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,109 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    MethodError("`transpose` not defined for `AbstractDataFrame`s. Try `permutedims` instead")
+
+"""
+    permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex,
```
Should that be `Union{Int, <:AbstractString}` then?

bkamins replied:
Yes, but only in the docsting, in the implementation `ColumnIndex` is OK

kescobo commented:
Alright I've addressed a bunch of these locally. Before I push stuff, I think there are a couple of open questions:

bkamins commented on docs/src/man/reshaping_and_pivoting.md:
````diff
@@ -380,3 +380,49 @@ julia> first(unstack(x, :Species, :vsum), 6)
+To turn an `AbstractDataFrame` on its side, use [`permutedims`](@ref).
+
+```jldoctest reshape
+julia> df1 = DataFrame(a=["x", "y"], b=[1.,2.], c=[3,4], d=[true,false])
````
apart from spaces (as we discussed earlier) - we typically write `1.0` not `1.` (here and elsewhere - for consistency)

kescobo commented:
Just noticed this:

```
permutedims(DataFrame(a=[:b,:c,:d]))
ERROR: ArgumentError: reducing over an empty collection is not allowed
```

Which I think need to be handled explicitly - seems like it should be `DataFrame(a=String[], b=[], c=[], d=[])` right?

kescobo commented:
Actually, enough changes were made to fix that that this is worth looking at I think...

kescobo commented on docs/src/man/reshaping_and_pivoting.md:
````diff
@@ -380,3 +380,49 @@ julia> first(unstack(x, :Species, :vsum), 6)
+Note also that the types of the other columns
+are the result of `promote_type` on _all_ the permuted columns.
+
+```jldoctest reshape
````
I was thinking that, since it's not recommended, but may be the result of many of this type of operation, it would be worth pointing out explicitly, no?

A thought I had was to include example code for fixing the problem, but the manual didn't seem like the right place for it.

bkamins replied:
I am OK with this example. I would just mention in the text above that a typical use case is for columns with homogenous eltype and we show here a case what happens when it is not the case.

kescobo replied:
Is this still live - @nalimilan do you still object, or are you OK with it?

bkamins replied:
I am OK with what we have now. Manual can be easily improved later if needed.

kescobo commented:
@bkamins Wow... you weren't kidding... 4 hours of cumulative CI time for each push! Passing now though :tada: 

I made a small tweak to language in docs for second example as [suggested here](https://github.com/JuliaData/DataFrames.jl/pull/2447#discussion_r501184842)

> Note that the first column (by default) of the original `df`
becomes the column names in the permuted result,
and the column names of the original become a new column.
Typically, this would be used on columns with homogenous element types,
since the element types of the other columns
are the result of `promote_type` on _all_ the permuted columns.

And removed `ColumnIndex` from docstring (though it's still in method signature).

Any thoughts on how I handled the single column case, or anything else I've missed?

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,113 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+                          makeunique::Bool=false)
+
+    if src_namescol isa Integer
+        1 <= src_namescol <= ncol(df) || throw(BoundsError("`src_namescol` doesn't exist"))
```
`BoundsError` does not take a string (see other `BoundsError` uses to see the syntax)

kescobo replied:
Can I do 

```
@boundscheck if src_namescol isa Integer
        1 <= src_namescol <= ncol(df) || throw(BoundsError(names(df), src_namescol))
    end
```

?

kescobo replied:
Now seeing examples that don't have `@boundscheck` - is that warranted here?

bkamins replied:
No - `@boundscheck` is only needed in performance sensitive code. This function is slow, so we should always do bounds checking.

nalimilan replied:
IIRC you should do something like `BoundsError(df, src_namescol)` as we don't want to print an error about `names(df)`.

kescobo replied:
This gives

kescobo replied:
I guess that's clearer than

```
julia> permutedims(DataFrame(x=["a","b","c"], y=rand(3), z=rand(3)), 4)
ERROR: BoundsError: attempt to access 3-element Vector{String} at index [4]
```

bkamins replied:
Normally we give the following error message:
```
julia> df = DataFrame(x=1)
1×1 DataFrame
│ Row │ x     │
│     │ Int64 │
├─────┼───────┤
│ 1   │ 1     │

julia> df[:, 2]
ERROR: BoundsError: attempt to access data frame with 1 column
  at index [2]
```
as I have suggested earlier - please look up the implementation there. We throw an error on `AbstractIndex` in order to generate a clean error message.

kescobo replied:
Apologies - I had looked at a couple of other examples, but evidently not the right one. It's not super obvious how stuff fits together, but I think I have it now:

```
julia> permutedims(DataFrame(x=["a","b","c"], y=rand(3), z=rand(3)), 4)
ERROR: BoundsError: attempt to access data frame with 3 columns at index [4]
```

bkamins replied:
Yes (and indeed DataFrames.jl is unfortunately complex, but we could not find a simpler design that would meet all the requirements people have)

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,113 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    if src_namescol isa Integer
+        1 <= src_namescol <= ncol(df) || throw(BoundsError("`src_namescol` doesn't exist"))
+    end
+    eltype(df[!, src_namescol]) <: SymbolOrString || throw(
```
I wuld also move `throw(` to the next line - now it looks a bit strange

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,113 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    df_permuted = DataFrame(dest_namescol => names(df_notsrc))
+
+    if ncol(df_notsrc) == 0
+        df_tmp = DataFrame(Dict(n=>[] for n in df[!, src_namescol]))
```
dict will reorder it is enough to write
```suggestion
        df_tmp = DataFrame((n=>[] for n in df[!, src_namescol])...)
```

kescobo replied:
Good catch! :+1:

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,113 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex;
+                          makeunique::Bool=false)
+    if src_namescol isa Integer
+        1 <= src_namescol <= ncol(df) || throw(BoundsError("`src_namescol` doesn't exist"))
```
same here

bkamins commented:
also please add NEWS.md entry when all else is done (it is best to do it as a last step as that file is changing often so this way you avoid having to constantly resolve merge conflicts).

kescobo commented:
> also please add NEWS.md

Can do - should I rebase onto `master` before doing this?

Anything else?

bkamins commented:
Yes, either rebase or merge (people often have problems when they rebase PRs that have many commits).

kescobo commented:
Alright, I did a merge and added NEWS.md entry. Anything else before I push hopefully last thing?

bkamins commented on docs/src/man/reshaping_and_pivoting.md:
````diff
@@ -380,3 +380,50 @@ julia> first(unstack(x, :Species, :vsum), 6)
+To turn an `AbstractDataFrame` on its side, use [`permutedims`](@ref).
+
+```jldoctest reshape
+julia> df1 = DataFrame(a=["x", "y"], b=[1.0, 2.0], c=[3, 4], d=[true,false])
````
```suggestion
julia> df1 = DataFrame(a=["x", "y"], b=[1.0, 2.0], c=[3, 4], d=[true, false])
```

kescobo replied:
Grrr... sneaky commas!

bkamins commented:
there are some unresolved suggestions from @nalimilan (+ I have added one minor comments still).

kescobo commented:
> there are some unresolved suggestions from @nalimilan

Do you mean [this one](there are some unresolved suggestions from @nalimilan)?

kescobo commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,109 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex,
+                dest_namescol::Union{Symbol,AbstractString};
+                makeunique::Bool=false)
+    permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex; makeunique::Bool=false)
+    permutedims(df::AbstractDataFrame; makeunique::Bool=false)
```
Oh this got hidden earlier when I was scrolling - fixing now...

kescobo replied:
Sorry, I'm not finding good examples of this - I've found one or two examples where the optional arguments don't have types (eg [stack](http://juliadata.github.io/DataFrames.jl/stable/lib/functions/#DataFrames.stack)). Only thing close is [show](http://juliadata.github.io/DataFrames.jl/stable/lib/functions/#Base.show), but this seems to treat the comma differently than the other examples (`stack(df::AbstractDataFrame, [measure_vars], [id_vars];` - comma outside brackets, separate brackets for each optional arg vs `show([io::IO,] df::AbstractDataFrame;` - comma inside bracket)

I also scanned through Base docs, and there's some inconsistency, eg `pop!(collection, key[, default])` (no space before the comma) vs `iterate(iter [, state])`, and `Float64(x [, mode::RoundingMode])`. Also, `match(r::Regex, s::AbstractString[, idx::Integer[, addopts]])` which has nested brackets, as does `reverse(v [, start=1 [, stop=length(v) ]] )

So, do you want:

option1:

```
    permutedims(df::AbstractDataFrame [, src_namescol::Union{Int, Symbol, <:AbstractString},
                dest_namescol::Union{Symbol, AbstractString}];
                makeunique::Bool=false)
```

option1b:

```
    permutedims(df::AbstractDataFrame[, src_namescol::Union{Int, Symbol, <:AbstractString},
                dest_namescol::Union{Symbol, AbstractString}];
                makeunique::Bool=false)
```

option2:

```
    permutedims(df::AbstractDataFrame, [src_namescol::Union{Int, Symbol, <:AbstractString},
                [dest_namescol::Union{Symbol, AbstractString}] ];
                makeunique::Bool=false)
```

or something else?

kescobo replied:
Actually, based on the `match` and `reverse` from base, I propose option3:

```
    permutedims(df::AbstractDataFrame [, src_namescol::Union{Int, Symbol, <:AbstractString}
                [, dest_namescol::Union{Symbol, AbstractString} ]];
                makeunique::Bool=false)
```

bkamins replied:
OK

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,113 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+                          makeunique::Bool=false)
+
+    if src_namescol isa Integer
+        1 <= src_namescol <= ncol(df) || throw(BoundsError(df, src_namescol))
```
but this `BoundsError` still looks not fixed (same below)

kescobo replied:
Yeah, I've got the fix locally. It's now `throw(BoundsError(index(df), src_namescol))`. Trying not to push small bunches of changes.

kescobo commented:
Still open

- [ ] Including example with `Any` columns in docs? [link](https://github.com/JuliaData/DataFrames.jl/pull/2447#discussion_r501992936)
- [ ] Docstring signature bracket placement? [link](https://github.com/JuliaData/DataFrames.jl/pull/2447#discussion_r500991243)
- [ ] Default `src_namescol` appropriate? [link](https://github.com/JuliaData/DataFrames.jl/pull/2447#discussion_r500992337)

bkamins commented:
For me the only thing to discuss it:

> Default `src_namescol` appropriate?

And I would stay with what @kescobo proposes, as doing just `permutedims(df)` is something that people would probably naturally want to do.

The only other option is to allow only:
1. a method with one argument (data frame) - then it would take the first column and retain its name
2. a method with three arguments (both input and output col) - with the idea that if someone needs to specify input column - it is probably not a first column so there is a higher probability that destination column name should be different

(this approach would be more restrictive - so we could always later loosen it if needed). I am not sure what is best (or maybe some other option?). But in general what is currently proposed in this PR is IMO acceptable.

kescobo commented:
A third approach I considered is to not have `src_namescol` as an argument, and _always_ pivot on the first column. A user can always use intending to put a different column first  (especially with newer syntax this is much easier). After all, `permutedims` on matrices doesn't allow putting a different column first.

Just a thought 🤷‍♂️

kescobo commented:
Test failure looks like a random bug, only happened on one build 

```
$ curl -A "$CURL_USER_AGENT" -sSf -L --retry 7 'https://julialang-s3.julialang.org/bin/linux/x86/1.0/julia-1.0-latest-linux-i686.tar.gz' | tar -C ~/julia -x -z --strip-components=1 -f -

curl: (56) GnuTLS recv error (-54): Error in the pull function.

gzip: stdin: unexpected end of file

tar: Unexpected EOF in archive

tar: Unexpected EOF in archive

tar: Error is not recoverable: exiting now

The command "curl -A "$CURL_USER_AGENT" -sSf -L --retry 7 'https://julialang-s3.julialang.org/bin/linux/x86/1.0/julia-1.0-latest-linux-i686.tar.gz' | tar -C ~/julia -x -z --strip-components=1 -f -" failed and exited with 2 during .
```

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,107 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+        m = permutedims(Matrix(df_notsrc))
+        df_tmp = rename!(DataFrame(Tables.table(m)), df[!, src_namescol], makeunique=makeunique)
+    end
+    return hcat!(df_permuted, df_tmp, copycols=false)
```
I think `makeunique` is missing here in `hcat!`. Can you please also add a test for this case (i.e. when it would be needed and if set to `false` it is an error and if set to `true` it is OK?)

kescobo replied:
Good catch - one above it too I think. 

I also went ahead and rebased onto current master. Did @nalimilan want to weigh in on the discussion about `src_namescol` default, or are we all set?

nalimilan replied:
Sounds OK. Though people may be surprised if they do e.g. `permutedims(DataFrame(x=[1, 2], y=[2, 3]))` and get an error about `src_colnames` as they didn't specify it. Could be worth having a specific error message for that method.

bkamins commented:
I have left one small comment regarding `makeunique`. Other than that it looks good to merge.

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,107 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+"""
+    permutedims(df::AbstractDataFrame [, src_namescol::Union{Int, Symbol, <:AbstractString}
+                    [, dest_namescol::Union{Symbol, AbstractString} ]];
+                    makeunique::Bool=false)
```
Align with `df`.

kescobo replied:
Ugh - a bunch of stuff seems like it got reverted after rebase somehow. Sorry about that

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,107 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+# Arguments
+- `df` : the `AbstractDataFrame`
+- `src_namescol` : the column that will become the new header.
+  This column eltype must be `<: Union{String, Symbol}`.
```
```suggestion
  This column's element type must be `AbstractString` or `Symbol`.
```

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,107 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+  with `_i` (`i` starting at 1 for the first duplicate).
+
+Note: The element types of columns in resulting `DataFrame`
+(other than the first column, w hich always has element type `String`)
```
```suggestion
(other than the first column, which always has element type `String`)
```

nalimilan commented on src/abstractdataframe/reshape.jl:
````diff
@@ -399,3 +399,107 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+# Examples
+
+```jldoctest
+julia> df1 = DataFrame(a=["x", "y"], b=[1.,2.], c=[3,4], d=[true,false])
````
Please add spaces after commas here and elsewhere for consistency.

nalimilan commented on src/abstractdataframe/reshape.jl:
````diff
@@ -399,3 +399,107 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+```
+"""
+function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex,
+                          dest_namescol::Union{Symbol, <:AbstractString};
````
```suggestion
                          dest_namescol::Union{Symbol, AbstractString};
```

bkamins commented:
> `src_colnames`

On a second thought maybe requiring to pass it (like the proposal `@nalimilan` made for `DataFrame` constructor with matrix today) would be a better solution. Writing `permutedims(df, 1)` is not that longer and does not require much more thought than `permutedims(df)`.

Especially that it now occurred to me that in the future we might want `permutedims(df)` to do something different - e.g. take all columns and generate dummy column names like `x1`, `x2`, etc.? What do you think?

kescobo commented:
> Especially that it now occurred to me that in the future we might want permutedims(df) to do something different - e.g. take all columns and generate dummy column names like x1, x2, etc.? What do you think?

You mean like 

```
permutedims(DataFrame(a=["x", "y, "z"], b=[1, 2, 3]) == DataFrame(x1=["x", 1], x2=["y", 2], x3=["z", 3])
```

Or should the column names also become a new column?

bkamins commented:
> Or should the column names also become a new column?

I am not sure - this is exactly what I was thinking of. The point is that we might change our minds what is good in the future so it is better to be conservative now (have a look at https://github.com/JuliaData/DataFrames.jl/pull/2464 where we started discussing things that worked for years and no one complained about them).

kescobo commented:
> The point is that we might change our minds what is good in the future so it is better to be conservative now (have a look at #2464 where we started discussing things that worked for years and no one complained about them).

TBH I don't 100% follow the relevance of that PR. Are you just saying that in the PR you're working on "fixing" things that have been around for years that people didn't complain about? 

Regardless, I definitely agree that being conservative is the right approach. But I'm not sure what the conservative version is in this case. Should we include a `keepcolnames::Bool` keyword to control this behavior? 

Just so I'm clear, the new behavior being proposed is:

- `permutedims(df, [colnames])`: permute all columns to rows, give dummy col names unless vector of names is provided. Open question about whether `names(df)` becomes a row or is lost.
- `permutedims(df, indexcol::ColumnIndex, [dest_colname])`: current behavior (that is, pivot on `indexcol`, option to give a different name to resulting first column)

bkamins commented:
> Are you just saying that in the PR you're working on "fixing" things that have been around for years that people didn't complain about?

No, I am saying that it is hard to get consensus what should be the behaviour of some functions.

bkamins commented:
> Just so I'm clear, the new behavior being proposed is:

What I am considering if we should allow `permutedims(df)` at all. Maybe at this stage it should be disallowed. We can decide what it means later.

kescobo commented:
Ahh, got it. That makes sense. Ok, will take another look tomorrow

bkamins commented on docs/src/man/reshaping_and_pivoting.md:
```diff
@@ -380,3 +380,50 @@ julia> first(unstack(x, :Species, :vsum), 6)
+│ 1   │ x      │ 1   │ 3     │ 1    │
+│ 2   │ y      │ two │ 4     │ 0    │
+
+julia> permutedims(df2)
```
this is outdated

bkamins commented on docs/src/man/reshaping_and_pivoting.md:
```diff
@@ -380,3 +380,50 @@ julia> first(unstack(x, :Species, :vsum), 6)
+│ 1   │ x      │ 1.0     │ 3     │ 1    │
+│ 2   │ y      │ 2.0     │ 4     │ 0    │
+
+julia> permutedims(df1)
```
this is outdated

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,106 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+│ 1   │ x      │ 1   │ 3     │ 1    │
+│ 2   │ y      │ two │ 4     │ 0    │
+
+julia> permutedims(df1, 1) # note the column types
```
I would reorder this. It is easier to read: 1) define a data frame, 2) transpose it. And then the same for the second one. Also in one of the examples add the third positional argument to show how it can be used.

kescobo replied:
Do you want that third example in the docs as well, or just the docstring?

bkamins replied:
I think 2 examples are enough - just one of them should show the 3 positional arguments form (probably best both in docstrings and in the manual). In general we try to cover all signatures in the examples where possible (sometimes it is not possible as there are too many signatures, but in this case it is not a problem I think)

bkamins reviewed (approved):
Looks good except for three places in the documentation where I left comments. Thank you!

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -417,7 +417,7 @@ with name specified by `dest_namescol`.
 - `df` : the `AbstractDataFrame`
 - `src_namescol` : the column that will become the new header.
-  This column's element type must be `AbstractString` or `Symbol`.
+  This column's element type must be `AbstractString`, `Symbol`, or `Int`.
```
really `Int` is allowed as `eltype`? I thought not.

kescobo replied:
Oh, I was thinking of `dest_namescol` :facepalm:

bkamins commented:
@nalimilan - I am OK to merge this.

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,106 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    MethodError("`transpose` not defined for `AbstractDataFrame`s. Try `permutedims` instead")
+
+"""
+    permutedims(df::AbstractDataFrame, src_namescol::Union{Int, Symbol, <:AbstractString}
```
```suggestion
    permutedims(df::AbstractDataFrame, src_namescol::Union{Int, Symbol, AbstractString},
```

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,106 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+                makeunique::Bool=false)
+
+Turn `df` on its side such that rows become columns
+and the column indexed by `src_namescol` becomes the names of new columns.
```
```suggestion
and values in the column indexed by `src_namescol` become the names of new columns.
```

nalimilan commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,106 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+    df_permuted = DataFrame(dest_namescol => names(df_notsrc))
+
+    if ncol(df_notsrc) == 0
+        df_tmp = DataFrame((n=>[] for n in df[!, src_namescol])..., makeunique=makeunique)
```
This will avoid specializing the constructor on the number of columns:
```suggestion
        df_tmp = DataFrame(AbstractVector[[] for _ in 1:nrow(df)], df[!, src_namescol],
                           makeunique=makeunique)
```

nalimilan reviewed (commented):
Thanks, looks good apart from a few small suggestions.

bkamins commented on src/abstractdataframe/reshape.jl:
```diff
@@ -399,3 +399,106 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
+
+"""
+    permutedims(df::AbstractDataFrame, src_namescol::Union{Int, Symbol, <:AbstractString}
+                [, dest_namescol::Union{Symbol, AbstractString}];
```
```suggestion
                [dest_namescol::Union{Symbol, AbstractString}];
```

bkamins commented:
I will merge the PR once the CI passes. Thank you!

kescobo commented:
Woo! Thank you both for thoughtful and detailed review! Really appreciate it :-)

bkamins commented:
Thank you!
